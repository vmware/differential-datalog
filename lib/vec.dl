/*
 * Functions for working with vectors (see basic vector operations defined in
 * `ddlog_std.dl`).
 */

/* Applies closure `f` to each element of the vector. */
function map(v: Vec<'A>, f: function('A): 'B): Vec<'B> {
    var res = vec_with_capacity(v.len());
    for (x in v) {
        res.push(f(x))
    };
    res
}

/* Applies closure `f` to each element of the vector and concatenates the
 * resulting vectors, returning a flat vector. */
function flatmap(v: Vec<'A>, f: function('A): Vec<'B>): Vec<'B> {
    var res = vec_empty();
    for (x in v) {
        res.append(f(x))
    };
    res
}

/* Sorts the vector with a key extraction function, but may not preserve the
 * order of equal elements. */
function sort_by(v: mut Vec<'A>, f: function('A): 'B) {
    vec_sort_by(v, f)
}

/* Returns the element that gives the minimum value from the specified function.
 * If several elements are equally minimum, the first element is returned.
 * If the vector is empty, `None` is returned. */
function min_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
    vec_min_by(v, f)
}

/* Returns the element that gives the maximum value from the specified function.
 * If several elements are equally maximum, the first element is returned.
 * If the vector is empty, `None` is returned. */
function max_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
    vec_max_by(v, f)
}

/* Returns the first element of the vector that satisfies predicate `f` or
 * `None` if none of the elements satisfy the predicate. */
function find(v: Vec<'A>, f: function('A): bool): Option<'A> {
    for (x in v) {
        if (f(x)) {
            return Some{x}
        }
    };
    None
}

/* Returns a vector containing only those elements in `v` that satisfy predicate `f`,
 * preserving the order of the elements in the original vector. */
function filter(v: Vec<'A>, f: function('A): bool): Vec<'A> {
    var res = vec_empty();
    for (x in v) {
        if (f(x)) {
            res.push(x)
        }
    };
    res
}

/* Both filters and maps the vector.
 *
 * Calls the closure on each element of the vector.  If the closure returns
 * `Some{element}`, then that element is returned. */
function filter_map(v: Vec<'A>, f: function('A): Option<'B>): Vec<'B> {
    var res = vec_empty();
    for (x in v) {
        match (f(x)) {
            None -> (),
            Some{y} -> res.push(y)
        }
    };
    res
}


/* Retains only the elements specified by the predicate.
 * In other words, remove all elements `e` such that `f(e)` returns `false`.
 * This method operates in place, visiting each element exactly once in the
 * original order, and preserves the order of the retained elements. */
function retain(v: mut Vec<'A>, f: function('A): bool) {
    var del: s64 = 0;
    var len = v.len();

    for (i in range(0, (len as isize) - 1, 1)) {
        var x = v.nth(i as usize).unwrap_or_default();
        if (not f(x)) {
            del = del + 1
        } else if (del > 0) {
            v.update_nth((i - del) as usize, x);
        }
    };
    if (del > 0) {
        v.truncate(len - (del as usize));
    }
}

/* Returns `true` iff all elements of the vector satisfy predicate `f`. */
function all(v: Vec<'A>, f: function('A): bool): bool {
    for (x in v) {
        if (not f(x)) {
            return false
        }
    };
    true
}

/* Returns `true` iff at least one element of the vector satisfies predicate `f`. */
function any(v: Vec<'A>, f: function('A): bool): bool {
    for (x in v) {
        if (f(x)) {
            return true
        }
    };
    false
}

/* Counts the number of elements in `v` that satisfy predicate `f`. */
function count(v: Vec<'A>, f: function('A): bool): usize {
    var cnt = 0;
    for (x in v) {
        if (f(x)) {
            cnt = cnt + 1
        }
    };
    cnt
}

/* Iterates over the vector, aggregating its contents using `f`.
 *
 * `f` - takes the previous value of the accumulator and the next element in the
 * vector and returns the new value of the accumulator.
 *
 * `initializer` - initial value of the accumulator. */
function fold(v: Vec<'A>, f: function('B, 'A): 'B, initializer: 'B): 'B {
    var res = initializer;
    for (x in v) {
        res = f(res, x)
    };
    res
}

extern function vec_sort_by(v: mut Vec<'A>, f: function('A): 'B)
extern function vec_min_by(v: Vec<'A>, f: function('A): 'B): Option<'A>
extern function vec_max_by(v: Vec<'A>, f: function('A): 'B): Option<'A>
