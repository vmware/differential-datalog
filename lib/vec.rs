use crate::closure::Closure;
use crate::ddlog_std;

pub fn vec_sort_by<A, B: Ord>(v: &mut ddlog_std::Vec<A>, f: &Box<dyn Closure<*const A, B>>) {
    v.x.sort_unstable_by_key(|x| f.call(x))
}

pub fn vec_min_by<A: Clone, B: Ord>(
    v: &ddlog_std::Vec<A>,
    f: &Box<dyn Closure<*const A, B>>,
) -> ddlog_std::Option<A> {
    ddlog_std::Option::from(v.x.iter().min_by_key(|x| f.call(*x)).map(|x| x.clone()))
}

pub fn vec_max_by<A: Clone, B: Ord>(
    v: &ddlog_std::Vec<A>,
    f: &Box<dyn Closure<*const A, B>>,
) -> ddlog_std::Option<A> {
    ddlog_std::Option::from(v.x.iter().max_by_key(|x| f.call(*x)).map(|x| x.clone()))
}
