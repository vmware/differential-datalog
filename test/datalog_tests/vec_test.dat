dump vec_test::IntVecTransforms;
dump vec_test::CompareClosures;
dump vec_test::VecFolds;
