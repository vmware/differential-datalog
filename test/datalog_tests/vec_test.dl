import vec

function int_vec(): Vec<s64> {
    range(-10, 10, 1)
}

output relation IntVecTransforms(description: string, v: Vec<s64>)

// filter.
IntVecTransforms("odds", int_vec().filter(|x| x % 2 != 0)).

function is_even(x: s64): bool {
    x % 2 == 0
}

IntVecTransforms("evens", int_vec().filter(is_even)).
// map.
IntVecTransforms("+100", int_vec().map(|x|x+100)).

// retain.
IntVecTransforms("(0,5]", {
                     var res = int_vec();
                     res.retain(|x|: bool { x > 0 and x <= 5 });
                     res
                 }).

IntVecTransforms(">3", {
                     var res = int_vec();
                     var threshold = 3;
                     res.retain(|x: s64| x > threshold);
                     res
                 }).

// find.
IntVecTransforms("not in [-5, 0, 7]", {
                     var res = int_vec();
                     var exclude = [-5, 0];
                     exclude.push(7);
                     res.retain(|x| is_none(exclude.find(|y| y==x )));
                     res
                 }).

// all.
IntVecTransforms("not in [-5, 0, 100] using 'all'", {
                     var res = int_vec();
                     var exclude = [-5, 0, 100];
                     res.retain(|x| exclude.all(|y| y != x)  );
                     res
                 }).

// any.
IntVecTransforms("not in [-5, 0, 100] using 'any'", {
                     var res = int_vec();
                     var exclude = [-5, 0, 100];
                     res.retain(|x| not exclude.any(|y| y == x)  );
                     res
                 }).

// sort_by.
IntVecTransforms("sort_by(identity)", {
                     var res = int_vec();
                     res.sort_by(|x|x);
                     res
                 }).

IntVecTransforms("sort_by(-x)", {
                     var res = int_vec();
                     res.sort_by(|x|-x);
                     res
                 }).

IntVecTransforms("sort_by(to_string)", {
                     var res = int_vec();
                     res.sort_by(to_string);
                     res
                 }).

// arg_min.
IntVecTransforms("arg_min(x*x)", vec_singleton(int_vec().arg_min(|x| x*x).unwrap_or_default())).

// arg_max.
IntVecTransforms("arg_max(to_string)", vec_singleton(int_vec().arg_max(to_string).unwrap_or_default())).

// filter_map.
IntVecTransforms("filter_map(if (x < 0){None} else {Some{x+1000}})", int_vec().filter_map(|x| if (x < 0) { None } else { Some{x + 1000} })).

// flatmap. 
IntVecTransforms("flatmap(|x| [x, x + delta1, x + delta2])",
                 {
                     var delta1 = -1000;
                     var delta2 = delta1 - 1000;
                     int_vec().flatmap(|x| [x, x + delta1, x + delta2])
                 }).

/* Use closures in relations. */
relation IntClosures(f: |s64|: s64)

relation Increments(i: s64)

Increments(100).
Increments(200).

IntClosures(|x| x + inc) :- Increments(inc).
IntClosures(|x| x*x).

typedef vec_to_vec_t = |Vec<s64>|: Vec<s64>

relation IntVecClosures(f: vec_to_vec_t)

IntVecClosures(|v|v.map(f)) :- IntClosures(f).
IntVecClosures(|v|v.map(ref_new).map(deref)).

IntVecTransforms("closure ${f}", f(int_vec())) :- IntVecClosures(f).

/* Comparing closures. */
output relation CompareClosures(c1: string, c2: string, equal: bool)

// fold.
CompareClosures("${c1}", "${c2}", c1 == c2) :-
    IntVecClosures(c1),
    IntVecClosures(c2).

output relation VecFolds(description: string, a: s64)

VecFolds("1+2+3+4+5", [1,2,3,4,5].fold(|a,x| a + x, 0)).
