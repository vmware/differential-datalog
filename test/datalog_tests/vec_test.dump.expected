vec_test::IntVecTransforms{.description = "(0,5]", .v = [1, 2, 3, 4, 5]}
vec_test::IntVecTransforms{.description = "+100", .v = [90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110]}
vec_test::IntVecTransforms{.description = ">3", .v = [4, 5, 6, 7, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "closure <closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .v = [-10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "closure <closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .v = [100, 81, 64, 49, 36, 25, 16, 9, 4, 1, 0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]}
vec_test::IntVecTransforms{.description = "closure <closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .v = [90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110]}
vec_test::IntVecTransforms{.description = "closure <closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .v = [190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210]}
vec_test::IntVecTransforms{.description = "evens", .v = [-10, -8, -6, -4, -2, 0, 2, 4, 6, 8, 10]}
vec_test::IntVecTransforms{.description = "filter_map(if (x < 0){None} else {Some{x+1000}})", .v = [1000, 1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1010]}
vec_test::IntVecTransforms{.description = "flatmap(|x| [x, x + delta1, x + delta2])", .v = [-10, -1010, -2010, -9, -1009, -2009, -8, -1008, -2008, -7, -1007, -2007, -6, -1006, -2006, -5, -1005, -2005, -4, -1004, -2004, -3, -1003, -2003, -2, -1002, -2002, -1, -1001, -2001, 0, -1000, -2000, 1, -999, -1999, 2, -998, -1998, 3, -997, -1997, 4, -996, -1996, 5, -995, -1995, 6, -994, -1994, 7, -993, -1993, 8, -992, -1992, 9, -991, -1991, 10, -990, -1990]}
vec_test::IntVecTransforms{.description = "max_by(to_string)", .v = [9]}
vec_test::IntVecTransforms{.description = "min_by(x*x)", .v = [0]}
vec_test::IntVecTransforms{.description = "not in [-5, 0, 100] uaing 'all'", .v = [-10, -9, -8, -7, -6, -4, -3, -2, -1, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "not in [-5, 0, 100] using 'any'", .v = [-10, -9, -8, -7, -6, -4, -3, -2, -1, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "not in [-5, 0, 7]", .v = [-10, -9, -8, -7, -6, -4, -3, -2, -1, 1, 2, 3, 4, 5, 6, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "odds", .v = [-9, -7, -5, -3, -1, 1, 3, 5, 7, 9]}
vec_test::IntVecTransforms{.description = "sort_by(-x)", .v = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5, -6, -7, -8, -9, -10]}
vec_test::IntVecTransforms{.description = "sort_by(identity)", .v = [-10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
vec_test::IntVecTransforms{.description = "sort_by(to_string)", .v = [-1, -10, -2, -3, -4, -5, -6, -7, -8, -9, 0, 1, 10, 2, 3, 4, 5, 6, 7, 8, 9]}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .equal = true}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .equal = true}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .equal = true}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map((vec::map(v, ddlog_std::ref_new)), ddlog_std::deref))}), captured_args: ()>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x * x)}), captured_args: ()>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 100>>", .equal = false}
vec_test::CompareClosures{.c1 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .c2 = "<closure: (function(v: ddlog_std::Vec<ddlog_std::s64>):ddlog_std::Vec<ddlog_std::s64>{(vec::map(v, f))}), captured_args: <closure: (function(x: ddlog_std::s64):ddlog_std::s64{(x + inc)}), captured_args: 200>>", .equal = true}
vec_test::VecFolds{.description = "1+2+3+4+5", .a = 15}
